File: scripts/work-ledger.py
#!/usr/bin/env python3
"""Append and summarize local work-ledger events."""

from __future__ import annotations

import argparse
import contextlib
import json
import sys
import tempfile
import uuid
from datetime import UTC, datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "scripts" / "lib"))

from agent_roster import ReceiptValidationError, validate_usage  # noqa: E402

DEFAULT_STORE = Path(".harness-kit/work/ledger.jsonl")
RECORD_TYPE = "work-ledger-event"
ACTIVE_STATUSES = {"active", "blocked"}
VALID_STATUSES = {"active", "blocked", "completed", "failed", "superseded"}
VALID_EVENT_TYPES = {
    "phase_started",
    "phase_completed",
    "blocker_added",
    "next_action_changed",
}


def now_iso() -> str:
    return datetime.now(UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def event_id() -> str:
    return f"work-{uuid.uuid4()}"


@contextlib.contextmanager
def file_lock(handle):
    try:
        import fcntl  # type: ignore
    except ImportError:
        yield
        return
    fcntl.flock(handle.fileno(), fcntl.LOCK_EX)
    try:
        yield
    finally:
        fcntl.flock(handle.fileno(), fcntl.LOCK_UN)


def build_event(args: argparse.Namespace) -> dict[str, object]:
    event: dict[str, object] = {
        "schema_version": 1,
        "record_type": RECORD_TYPE,
        "event_id": event_id(),
        "created_at": now_iso(),
        "event_type": args.event_type,
        "work_id": args.work_id,
        "parent_work_id": args.parent_work_id,
        "backlog_ref": args.backlog,
        "branch": args.branch,
        "owning_skill": args.owning_skill,
        "phase": args.phase,
        "evidence_refs": args.evidence_ref,
        "blockers": args.blocker,
        "spawned_agents": args.spawned_agent,
        "trace_refs": args.trace_ref,
        "next_action": args.next_action,
        "status": args.status,
    }
    if args.usage is not None:
        event["usage"] = args.usage
    return event


def append_event(store: Path, event: dict[str, object]) -> None:
    store.parent.mkdir(parents=True, exist_ok=True)
    with store.open("a", encoding="utf-8") as handle:
        with file_lock(handle):
            handle.write(json.dumps(event, sort_keys=True, separators=(",", ":")))
            handle.write("\n")


def read_events(store: Path) -> list[dict[str, object]]:
    if not store.exists():
        return []
    events: list[dict[str, object]] = []
    for lineno, line in enumerate(store.read_text(encoding="utf-8").splitlines(), start=1):
        if not line.strip():
            continue
        event = json.loads(line)
        if not isinstance(event, dict):
            raise ValueError(f"{store}:{lineno}: event must be a JSON object")
        events.append(event)
    return events


def latest_by_work(events: list[dict[str, object]]) -> dict[str, dict[str, object]]:
    latest: dict[str, dict[str, object]] = {}
    for event in events:
        latest[str(event["work_id"])] = event
    return latest


def format_list(values: object) -> str:
    if not values:
        return "none"
    if isinstance(values, list):
        return ", ".join(str(value) for value in values) if values else "none"
    return str(values)


def summary_text(events: list[dict[str, object]]) -> str:
    active = [
        event
        for event in latest_by_work(events).values()
        if str(event.get("status", "")) in ACTIVE_STATUSES
    ]
    if not active:
        return "No active work ledger entries."

    lines = ["Work ledger"]
    for event in sorted(active, key=lambda row: str(row.get("created_at", ""))):
        evidence_refs = event.get("evidence_refs")
        latest_evidence = "none"
        if isinstance(evidence_refs, list) and evidence_refs:
            latest_evidence = str(evidence_refs[-1])
        lines.extend(
            [
                f"- work_id: {event.get('work_id', '')}",
                f"  branch: {event.get('branch', '')}",
                f"  backlog: {event.get('backlog_ref', '')}",
                f"  event_type: {event.get('event_type', '')}",
                f"  owning_skill: {event.get('owning_skill', '')}",
                f"  phase: {event.get('phase', '')}",
                f"  status: {event.get('status', '')}",
                f"  latest_evidence: {latest_evidence}",
                f"  blockers: {format_list(event.get('blockers'))}",
                f"  spawned_agents: {format_list(event.get('spawned_agents'))}",
                f"  trace_refs: {format_list(event.get('trace_refs'))}",
                f"  next_action: {event.get('next_action', '')}",
            ]
        )
    return "\n".join(lines)


def append_command(args: argparse.Namespace) -> int:
    args.usage = _parse_usage_json(args.usage_json)
    event = build_event(args)
    append_event(args.store, event)
    print(json.dumps({"event_id": event["event_id"], "store": str(args.store)}, sort_keys=True))
    return 0


def summary_command(args: argparse.Namespace) -> int:
    try:
        events = read_events(args.store)
    except (json.JSONDecodeError, ValueError) as error:
        print(f"work-ledger: {error}", file=sys.stderr)
        return 2
    print(summary_text(events))
    return 0


def self_test() -> int:
    with tempfile.TemporaryDirectory() as tmp:
        store = Path(tmp) / "ledger.jsonl"
        args = argparse.Namespace(
            store=store,
            event_type="phase_started",
            work_id="058",
            parent_work_id="",
            backlog="058",
            branch="deliver/058-work-ledger-mission-control",
            owning_skill="deliver",
            phase="review",
            evidence_ref=[".harness-kit/traces/delegations.jsonl#abc"],
            blocker=["waiting for critic"],
            spawned_agent=["grok-build:critic"],
            trace_ref=[".harness-kit/traces/work-records.jsonl#trace-abc"],
            next_action="address critic output",
            status="active",
            usage={
                "input_tokens": 100,
                "output_tokens": 25,
                "total_tokens": 125,
                "cost_usd": 0.01,
                "cost_source": "manual",
            },
        )
        append_event(store, build_event(args))
        text = summary_text(read_events(store))
        assert "backlog: 058" in text
        assert "phase: review" in text
        assert "trace_refs: .harness-kit/traces/work-records.jsonl#trace-abc" in text

        args.phase = "done"
        args.event_type = "phase_completed"
        args.status = "completed"
        args.blocker = []
        args.spawned_agent = []
        args.trace_ref = []
        args.evidence_ref = []
        args.next_action = "none"
        args.usage = None
        append_event(store, build_event(args))
        assert summary_text(read_events(store)) == "No active work ledger entries."
    print("work-ledger self-test ok")
    return 0


def _parse_usage_json(value: str | None) -> dict[str, object] | None:
    if not value:
        return None
    try:
        usage = json.loads(value)
    except json.JSONDecodeError as error:
        raise SystemExit(f"work-ledger: invalid --usage-json: {error}") from error
    try:
        validate_usage(usage)
    except ReceiptValidationError as error:
        raise SystemExit(f"work-ledger: invalid --usage-json: {error}") from error
    return usage


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--self-test", action="store_true")
    subparsers = parser.add_subparsers(dest="command")

    append = subparsers.add_parser("append", help="append a phase transition event")
    append.add_argument("--store", type=Path, default=DEFAULT_STORE)
    append.add_argument("--event-type", choices=sorted(VALID_EVENT_TYPES), required=True)
    append.add_argument("--work-id", required=True)
    append.add_argument("--parent-work-id", default="")
    append.add_argument("--backlog", required=True)
    append.add_argument("--branch", required=True)
    append.add_argument("--owning-skill", required=True)
    append.add_argument("--phase", required=True)
    append.add_argument("--evidence-ref", action="append", default=[])
    append.add_argument("--blocker", action="append", default=[])
    append.add_argument("--spawned-agent", action="append", default=[])
    append.add_argument("--trace-ref", action="append", default=[])
    append.add_argument("--next-action", required=True)
    append.add_argument("--status", choices=sorted(VALID_STATUSES), default="active")
    append.add_argument("--usage-json", default=None)
    append.set_defaults(func=append_command)

    summary = subparsers.add_parser("summary", help="print active work summary")
    summary.add_argument("--store", type=Path, default=DEFAULT_STORE)
    summary.set_defaults(func=summary_command)
    return parser


def main(argv: list[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)
    if args.self_test:
        return self_test()
    if not hasattr(args, "func"):
        parser.error("choose a command or --self-test")
    return args.func(args)


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/harness-kit-checks/src/work_ledger.rs
use std::fs::{self, File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use chrono::{SecondsFormat, Utc};
use serde_json::{Map as JsonMap, Value as JsonValue, json};
use uuid::Uuid;

use crate::source_refs;

pub const DEFAULT_STORE: &str = ".harness-kit/work/ledger.jsonl";
const RECORD_TYPE: &str = "work-ledger-event";

const ACTIVE_STATUSES: &[&str] = &["active", "blocked"];
const VALID_STATUSES: &[&str] = &["active", "blocked", "completed", "failed", "superseded"];
const VALID_COST_SOURCES: &[&str] = &["provider_reported", "estimated", "manual", "unknown"];
const VALID_EVENT_TYPES: &[&str] = &[
    "phase_started",
    "phase_completed",
    "blocker_added",
    "next_action_changed",
];

#[derive(Debug, Clone, PartialEq)]
pub struct AppendOptions {
    pub store: PathBuf,
    pub event_type: String,
    pub work_id: String,
    pub parent_work_id: String,
    pub backlog: String,
    pub branch: String,
    pub owning_skill: String,
    pub phase: String,
    pub evidence_refs: Vec<String>,
    pub blockers: Vec<String>,
    pub spawned_agents: Vec<String>,
    pub trace_refs: Vec<String>,
    pub next_action: String,
    pub status: String,
    pub usage: Option<JsonValue>,
    pub work_source_refs: Vec<JsonValue>,
}

pub fn default_store() -> PathBuf {
    PathBuf::from(DEFAULT_STORE)
}

pub fn build_event(options: &AppendOptions) -> Result<JsonValue> {
    validate_enum("event_type", &options.event_type, VALID_EVENT_TYPES)?;
    validate_enum("status", &options.status, VALID_STATUSES)?;
    if let Some(usage) = &options.usage {
        validate_usage(usage).context("invalid --usage-json")?;
    }
    source_refs::validate_refs(&options.work_source_refs, Some(&options.backlog))
        .context("invalid work source refs")?;

    let mut event = JsonMap::new();
    event.insert("backlog_ref".to_string(), json!(options.backlog));
    event.insert("blockers".to_string(), json!(options.blockers));
    event.insert("branch".to_string(), json!(options.branch));
    event.insert(
        "created_at".to_string(),
        json!(Utc::now().to_rfc3339_opts(SecondsFormat::Secs, true)),
    );
    event.insert(
        "event_id".to_string(),
        json!(format!("work-{}", Uuid::new_v4())),
    );
    event.insert("event_type".to_string(), json!(options.event_type));
    event.insert("evidence_refs".to_string(), json!(options.evidence_refs));
    event.insert("next_action".to_string(), json!(options.next_action));
    event.insert("owning_skill".to_string(), json!(options.owning_skill));
    event.insert("parent_work_id".to_string(), json!(options.parent_work_id));
    event.insert("phase".to_string(), json!(options.phase));
    event.insert("record_type".to_string(), json!(RECORD_TYPE));
    event.insert("schema_version".to_string(), json!(1));
    event.insert("spawned_agents".to_string(), json!(options.spawned_agents));
    event.insert("status".to_string(), json!(options.status));
    event.insert("trace_refs".to_string(), json!(options.trace_refs));
    if !options.work_source_refs.is_empty() {
        event.insert(
            source_refs::FIELD.to_string(),
            json!(options.work_source_refs),
        );
    }
    if let Some(usage) = &options.usage {
        event.insert("usage".to_string(), usage.clone());
    }
    event.insert("work_id".to_string(), json!(options.work_id));
    Ok(JsonValue::Object(event))
}

pub fn append_event(store: &Path, event: &JsonValue) -> Result<()> {
    if let Some(parent) = store.parent() {
        fs::create_dir_all(parent)
            .with_context(|| format!("failed to create {}", parent.display()))?;
    }
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(store)
        .with_context(|| format!("failed to open {}", store.display()))?;
    lock_file(&file)?;
    let write_result = writeln!(
        file,
        "{}",
        serde_json::to_string(event).context("failed to serialize work-ledger event")?
    );
    let unlock_result = unlock_file(&file);
    write_result.with_context(|| format!("failed to append {}", store.display()))?;
    unlock_result?;
    Ok(())
}

pub fn append(options: &AppendOptions) -> Result<JsonValue> {
    let event = build_event(options)?;
    append_event(&options.store, &event)?;
    Ok(json!({
        "event_id": event.get("event_id").and_then(JsonValue::as_str).unwrap_or_default(),
        "store": options.store.to_string_lossy(),
    }))
}

pub fn read_events(store: &Path) -> Result<Vec<JsonMap<String, JsonValue>>> {
    if !store.exists() {
        return Ok(Vec::new());
    }
    let file = File::open(store).with_context(|| format!("failed to read {}", store.display()))?;
    let mut events = Vec::new();
    for (index, line) in BufReader::new(file).lines().enumerate() {
        let line_number = index + 1;
        let line = line.with_context(|| format!("{}:{line_number}", store.display()))?;
        if line.trim().is_empty() {
            continue;
        }
        let event: JsonValue = serde_json::from_str(&line)
            .with_context(|| format!("{}:{line_number}", store.display()))?;
        let JsonValue::Object(object) = event else {
            bail!(
                "{}:{line_number}: event must be a JSON object",
                store.display()
            );
        };
        events.push(object);
    }
    Ok(events)
}

pub fn summary(store: &Path) -> Result<String> {
    summary_text(&read_events(store)?)
}

pub fn summary_text(events: &[JsonMap<String, JsonValue>]) -> Result<String> {
    let mut latest: JsonMap<String, JsonValue> = JsonMap::new();
    for event in events {
        let work_id = event
            .get("work_id")
            .ok_or_else(|| anyhow::anyhow!("event missing work_id"))?
            .to_string();
        latest.insert(work_id, JsonValue::Object(event.clone()));
    }

    let mut active: Vec<JsonMap<String, JsonValue>> = latest
        .values()
        .filter_map(JsonValue::as_object)
        .filter(|event| {
            event
                .get("status")
                .and_then(JsonValue::as_str)
                .is_some_and(|status| ACTIVE_STATUSES.contains(&status))
        })
        .cloned()
        .collect();
    if active.is_empty() {
        return Ok("No active work ledger entries.".to_string());
    }
    active.sort_by_key(|event| string_field(event, "created_at"));

    let mut lines = vec!["Work ledger".to_string()];
    for event in active {
        let latest_evidence = event
            .get("evidence_refs")
            .and_then(JsonValue::as_array)
            .and_then(|values| values.last())
            .map(format_scalar)
            .unwrap_or_else(|| "none".to_string());
        lines.extend([
            format!("- work_id: {}", string_field(&event, "work_id")),
            format!("  branch: {}", string_field(&event, "branch")),
            format!("  backlog: {}", string_field(&event, "backlog_ref")),
            format!("  event_type: {}", string_field(&event, "event_type")),
            format!("  owning_skill: {}", string_field(&event, "owning_skill")),
            format!("  phase: {}", string_field(&event, "phase")),
            format!("  status: {}", string_field(&event, "status")),
            format!("  latest_evidence: {latest_evidence}"),
            format!("  blockers: {}", format_list(event.get("blockers"))),
            format!(
                "  spawned_agents: {}",
                format_list(event.get("spawned_agents"))
            ),
            format!("  trace_refs: {}", format_list(event.get("trace_refs"))),
            format!("  next_action: {}", string_field(&event, "next_action")),
        ]);
    }
    Ok(lines.join("\n"))
}

pub fn parse_usage_json(value: Option<&str>) -> Result<Option<JsonValue>> {
    let Some(value) = value else {
        return Ok(None);
    };
    let usage: JsonValue =
        serde_json::from_str(value).with_context(|| format!("invalid --usage-json: {value}"))?;
    if usage.is_null() {
        return Ok(None);
    }
    validate_usage(&usage).context("invalid --usage-json")?;
    Ok(Some(usage))
}

pub fn validate_usage(usage: &JsonValue) -> Result<()> {
    let Some(object) = usage.as_object() else {
        bail!("usage must be an object or null.");
    };
    for key in object.keys() {
        if ![
            "input_tokens",
            "output_tokens",
            "total_tokens",
            "cost_usd",
            "cost_source",
        ]
        .contains(&key.as_str())
        {
            bail!("usage has unknown fields: {key}");
        }
    }
    for token_field in ["input_tokens", "output_tokens", "total_tokens"] {
        if let Some(value) = object.get(token_field)
            && !(value.is_null() || value.as_u64().is_some())
        {
            bail!("usage {token_field} must be a non-negative integer or null.");
        }
    }
    if let Some(value) = object.get("cost_usd")
        && !(value.is_null() || value.as_f64().is_some_and(|number| number >= 0.0))
    {
        bail!("usage cost_usd must be a non-negative number or null.");
    }
    if let Some(value) = object.get("cost_source")
        && !(value.is_null()
            || value
                .as_str()
                .is_some_and(|source| VALID_COST_SOURCES.contains(&source)))
    {
        bail!("usage cost_source is invalid.");
    }
    if object.get("cost_usd").is_some_and(|value| !value.is_null())
        && object
            .get("cost_source")
            .is_none_or(|value| value.is_null())
    {
        bail!("usage cost_source is required when cost_usd is known.");
    }
    Ok(())
}

pub fn self_test() -> Result<()> {
    let temp = tempfile::tempdir().context("failed to create temporary directory")?;
    let store = temp.path().join("ledger.jsonl");
    let mut options = AppendOptions {
        store: store.clone(),
        event_type: "phase_started".to_string(),
        work_id: "058".to_string(),
        parent_work_id: String::new(),
        backlog: "058".to_string(),
        branch: "deliver/058-work-ledger-mission-control".to_string(),
        owning_skill: "deliver".to_string(),
        phase: "review".to_string(),
        evidence_refs: vec![".harness-kit/traces/delegations.jsonl#abc".to_string()],
        blockers: vec!["waiting for critic".to_string()],
        spawned_agents: vec!["grok-build:critic".to_string()],
        trace_refs: vec![".harness-kit/traces/work-records.jsonl#trace-abc".to_string()],
        next_action: "address critic output".to_string(),
        status: "active".to_string(),
        usage: Some(json!({
            "input_tokens": 100,
            "output_tokens": 25,
            "total_tokens": 125,
            "cost_usd": 0.01,
            "cost_source": "manual",
        })),
        work_source_refs: Vec::new(),
    };
    append_event(&store, &build_event(&options)?)?;
    let text = summary(&store)?;
    assert!(text.contains("backlog: 058"));
    assert!(text.contains("phase: review"));
    assert!(text.contains("trace_refs: .harness-kit/traces/work-records.jsonl#trace-abc"));

    options.phase = "done".to_string();
    options.event_type = "phase_completed".to_string();
    options.status = "completed".to_string();
    options.blockers.clear();
    options.spawned_agents.clear();
    options.trace_refs.clear();
    options.evidence_refs.clear();
    options.next_action = "none".to_string();
    options.usage = None;
    append_event(&store, &build_event(&options)?)?;
    assert_eq!(summary(&store)?, "No active work ledger entries.");
    Ok(())
}

fn validate_enum(name: &str, value: &str, valid: &[&str]) -> Result<()> {
    if valid.contains(&value) {
        Ok(())
    } else {
        bail!("{name} must be one of: {}", valid.join(", "))
    }
}

fn string_field(event: &JsonMap<String, JsonValue>, key: &str) -> String {
    event.get(key).map(format_scalar).unwrap_or_default()
}

fn format_list(value: Option<&JsonValue>) -> String {
    let Some(value) = value else {
        return "none".to_string();
    };
    if let Some(values) = value.as_array() {
        if values.is_empty() {
            "none".to_string()
        } else {
            values
                .iter()
                .map(format_scalar)
                .collect::<Vec<_>>()
                .join(", ")
        }
    } else {
        format_scalar(value)
    }
}

fn format_scalar(value: &JsonValue) -> String {
    if let Some(value) = value.as_str() {
        value.to_string()
    } else {
        value.to_string()
    }
}

#[cfg(unix)]
fn lock_file(file: &File) -> Result<()> {
    let result = unsafe { libc::flock(std::os::fd::AsRawFd::as_raw_fd(file), libc::LOCK_EX) };
    if result == 0 {
        Ok(())
    } else {
        Err(std::io::Error::last_os_error()).context("failed to lock work-ledger store")
    }
}

#[cfg(unix)]
fn unlock_file(file: &File) -> Result<()> {
    let result = unsafe { libc::flock(std::os::fd::AsRawFd::as_raw_fd(file), libc::LOCK_UN) };
    if result == 0 {
        Ok(())
    } else {
        Err(std::io::Error::last_os_error()).context("failed to unlock work-ledger store")
    }
}

#[cfg(not(unix))]
fn lock_file(_file: &File) -> Result<()> {
    Ok(())
}

#[cfg(not(unix))]
fn unlock_file(_file: &File) -> Result<()> {
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_options(store: PathBuf) -> AppendOptions {
        AppendOptions {
            store,
            event_type: "phase_started".to_string(),
            work_id: "058".to_string(),
            parent_work_id: "flywheel-058".to_string(),
            backlog: "058".to_string(),
            branch: "deliver/058-work-ledger-mission-control".to_string(),
            owning_skill: "deliver".to_string(),
            phase: "review".to_string(),
            evidence_refs: vec![".harness-kit/traces/delegations.jsonl#abc".to_string()],
            blockers: vec!["waiting for critic".to_string()],
            spawned_agents: vec!["grok-build:critic".to_string()],
            trace_refs: vec![".harness-kit/traces/work-records.jsonl#trace-abc".to_string()],
            next_action: "address critic output".to_string(),
            status: "active".to_string(),
            usage: None,
            work_source_refs: Vec::new(),
        }
    }

    #[test]
    fn work_ledger_appends_compact_event_and_summary_matches_shell_contract() -> Result<()> {
        let temp = tempfile::tempdir()?;
        let store = temp.path().join("nested/ledger.jsonl");
        let options = sample_options(store.clone());

        let receipt = append(&options)?;
        assert_eq!(receipt["store"].as_str().unwrap(), store.to_string_lossy());
        assert!(receipt["event_id"].as_str().unwrap().starts_with("work-"));

        let rows: Vec<JsonValue> = fs::read_to_string(&store)?
            .lines()
            .map(serde_json::from_str)
            .collect::<Result<_, _>>()?;
        let raw = fs::read_to_string(&store)?;
        assert!(raw.starts_with("{\"backlog_ref\":"));
        assert_eq!(rows.len(), 1);
        let row = rows[0].as_object().unwrap();
        assert_eq!(row["schema_version"], 1);
        assert_eq!(row["record_type"], RECORD_TYPE);
        assert_eq!(row["event_type"], "phase_started");
        assert_eq!(row["work_id"], "058");
        assert_eq!(row["parent_work_id"], "flywheel-058");
        assert_eq!(row["backlog_ref"], "058");
        assert_eq!(row["branch"], "deliver/058-work-ledger-mission-control");
        assert_eq!(row["owning_skill"], "deliver");
        assert_eq!(row["phase"], "review");
        assert_eq!(
            row["evidence_refs"],
            json!([".harness-kit/traces/delegations.jsonl#abc"])
        );
        assert_eq!(row["blockers"], json!(["waiting for critic"]));
        assert_eq!(row["spawned_agents"], json!(["grok-build:critic"]));
        assert_eq!(
            row["trace_refs"],
            json!([".harness-kit/traces/work-records.jsonl#trace-abc"])
        );
        assert_eq!(row["next_action"], "address critic output");
        assert_eq!(row["status"], "active");
        assert!(row["created_at"].as_str().unwrap().ends_with('Z'));

        let text = summary(&store)?;
        assert!(text.contains("branch: deliver/058-work-ledger-mission-control"));
        assert!(text.contains("backlog: 058"));
        assert!(text.contains("event_type: phase_started"));
        assert!(text.contains("phase: review"));
        assert!(text.contains("latest_evidence: .harness-kit/traces/delegations.jsonl#abc"));
        assert!(text.contains("blockers: waiting for critic"));
        assert!(text.contains("spawned_agents: grok-build:critic"));
        assert!(text.contains("trace_refs: .harness-kit/traces/work-records.jsonl#trace-abc"));
        assert!(text.contains("next_action: address critic output"));
        Ok(())
    }

    #[test]
    fn work_ledger_records_optional_work_source_refs() -> Result<()> {
        let temp = tempfile::tempdir()?;
        let mut options = sample_options(temp.path().join("ledger.jsonl"));
        options.work_source_refs = vec![json!({
            "role": "backlog",
            "kind": "local_backlog",
            "id": "058",
            "uri": "backlog.d/058-work-ledger-mission-control.md",
            "closure": {"mode": "local_archive"}
        })];

        let row = build_event(&options)?;

        assert_eq!(
            row["work_source_refs"][0]["uri"],
            "backlog.d/058-work-ledger-mission-control.md"
        );
        Ok(())
    }

    #[test]
    fn work_ledger_rejects_local_backlog_ref_mismatch() {
        let temp = tempfile::tempdir().unwrap();
        let mut options = sample_options(temp.path().join("ledger.jsonl"));
        options.work_source_refs = vec![json!({
            "role": "backlog",
            "kind": "local_backlog",
            "id": "059"
        })];

        let error = build_event(&options).unwrap_err().to_string();
        assert!(error.contains("work source refs"));
    }

    #[test]
    fn work_ledger_latest_terminal_status_hides_active_work() -> Result<()> {
        let temp = tempfile::tempdir()?;
        let store = temp.path().join("ledger.jsonl");
        let mut options = sample_options(store.clone());
        append(&options)?;

        options.event_type = "phase_completed".to_string();
        options.phase = "done".to_string();
        options.status = "completed".to_string();
        options.evidence_refs.clear();
        options.blockers.clear();
        options.spawned_agents.clear();
        options.trace_refs.clear();
        options.next_action = "none".to_string();
        append(&options)?;

        assert_eq!(summary(&store)?, "No active work ledger entries.");
        Ok(())
    }

    #[test]
    fn work_ledger_missing_store_summarizes_empty() -> Result<()> {
        let temp = tempfile::tempdir()?;
        assert_eq!(
            summary(&temp.path().join("missing.jsonl"))?,
            "No active work ledger entries."
        );
        Ok(())
    }

    #[test]
    fn work_ledger_rejects_bad_jsonl_and_non_object_rows() -> Result<()> {
        let temp = tempfile::tempdir()?;
        let store = temp.path().join("ledger.jsonl");
        fs::write(&store, "[]\n")?;
        let error = read_events(&store).unwrap_err().to_string();
        assert!(error.contains("event must be a JSON object"));

        fs::write(&store, "{bad\n")?;
        let error = read_events(&store).unwrap_err().to_string();
        assert!(error.contains("ledger.jsonl:1"));
        Ok(())
    }

    #[test]
    fn work_ledger_usage_validation_matches_python_contract() {
        assert!(validate_usage(&json!({"cost_usd": 0.01, "cost_source": "manual"})).is_ok());
        assert!(validate_usage(&json!({"input_tokens": true})).is_err());
        assert!(validate_usage(&json!({"input_tokens": -1})).is_err());
        assert!(validate_usage(&json!({"cost_usd": 0.01})).is_err());
        assert!(validate_usage(&json!({"cost_source": "bogus"})).is_err());
        assert!(validate_usage(&json!({"extra": 1})).is_err());
        assert!(validate_usage(&json!(null)).is_err());
        assert_eq!(parse_usage_json(Some("null")).unwrap(), None);
    }

    #[test]
    fn work_ledger_self_test_covers_embedded_contract() -> Result<()> {
        self_test()
    }
}
